### ISSUE TITLE
Merge `jsonapi-serializer` with `jsonapi.rb` ?

### ISSUE THREAD

**Issue by stas** [jsonapi-serializer/jsonapi-serializer#144]
As I'm moving towards a major refactoring of this project I'd like to lunch for a community call.

A bit of a backstory, I'm currently, unofficially, maintaining 3 popular-ish JSONAPI related projects:
 * this one, aka `jsonapi-serializer`
 * https://github.com/stas/jsonapi.rb - which provides support for a (almost) full coverage of the JSONAPI spec (aka, filters, pagination, errors, and so on)
 * https://github.com/jsonapi-rb/jsonapi-rspec - which offers RSpec matchers to help us test JSONAPI apps

While the last project is _very well thank you_ on it's own, the first two, historically and lately are quite dependent on by the other.

In order to make our life a bit more simple, I'm considering merging most if not all of the `jsonapi.rb` into `jsonapi-serializer`. And although the main goal is to not introduce any new dependencies and keep things as _vanilla_ as possible, this means more code to maintain (~500). On the bright side of things, the users of `jsonapi-serializer` can benefit from goodness like:
 * error handling and support
 * deserialization
 * pagination
 * filtering (with a soft-dependency for `ransack`)
 * and finally rails integration

Both projects are modular and are designer to let you use just what you want.

I'll be happy to hear your thoughts in regards to this. Thank you! :bow:

**Comment by ChrisHampel** [comment]
Part of why I used fast_jsonapi(in the process of converting over to this library from it) in the first place was that it was purely a serializer. We have systems that handle: filtering, pagination, deserialization, and errors. Thus, having a library that is just a serializer was preferable.

**Comment by jopotts** [comment]
I agree with @ChrisHampel - we also chose jsonapi-serializer (formerly fast-jsonapi) because it was simple and allowed us to implement the other parts of the stack ourselves. The serialization part is universal, whereas the other parts are normally very particular to the system. So if there's a consolidation, then I'd favour that happening as a new super-gem that uses this as a sub-gem.

Smaller gems composed into a bigger all-encompassing one would be better (similar to how rails does it).

And as always, thank you so much for all your super hard work @stas 🏅

**Comment by stas** [comment]
@jopotts @ChrisHampel notes taken. 

Still I'd be happy to hear your thoughts regarding the serialization features (ex. error serializers and deserialization). These were brought up recently and commonly in the `fast_jsonapi`.

**Comment by ziggythehamster** [comment]
I'm with the above users. We actually are going to be using `jsonapi.rb`, but with many things bypassed/overridden due to backcompat with our previous in-house JSON:API support and preferences (e.g., I always want to render pagination links, if pagination is possible, even if the field is null because there's no page prev/next, to be explicit, even though that's not required). I would prefer having the gem separate, as it would prevent accidental coupling between the two.

**Comment by ekampp** [comment]
Also, jsonapi.rb doesn't really have great support for databases that don't have full support for ActiveRecord. 

I used it on a major project that used Neo4j, and we had a lot of problems with it, and we ended up having to duck type a lot of jsonapi.rb's code, as well as selectively only load specific parts. 

I have used this gem (and its predecessor) for many years, and I have been delighted with it. I would certainly move away to something else if this merge happened, and I would be saddened to do so.

**Comment by stas** [comment]
@ekampp would be great if you could share the issues you encountered with Neo4j, or not only :bow: 

As for the merge, again, the question is more about whether there are parts that are worth being merged due to the previous community requests:
https://github.com/Netflix/fast_jsonapi/issues

Thanks again for your feedback!

**Comment by ekampp** [comment]
Hi, @stas, thanks for inquiring about that. 

The main problem is that neo4j doesn't use tables. So all the parts about table prefixing had to be overridden. 

Because the field selector is also different, we had to override a large part of jsonapi.rb's implementation of Ransack. This means overriding sections relating to both filtering, sorting, and sparse fieldsets. 

I can totally understand why people want to merge this with jsonapi.rb. I would have agreed had I ever only used bog-standard ActiveRecord. 

But the fact is that this gem, jsonapi-serializer is so brilliantly simple, has no advanced dependencies, is fast, and does the thing it claims to do really well. 

My years of experience have led me to the conclusion that composition is better than configuration. And I believe that merging jsonapi-serializer is an attempt at making a _magic bullet_™️.

**Comment by stas** [comment]
> The main problem is that neo4j doesn't use tables. So all the parts about table prefixing had to be overridden.
> 
> Because the field selector is also different, we had to override a large part of jsonapi.rb's implementation of Ransack. This means overriding sections relating to both filtering, sorting, and sparse fieldsets.

Hmm, the reason `jsonapi.rb` is composed of modules, is exactly for this reason. If you're ok sharing your work for Neo4j on `JSONAPI::Filtering`, I'll be happy to look into merging it!


As for _magic bullet_, that's not the goal, I can totally take the responsibility for that!

Tbh, my main desire would be to have error handling as part of this gem, since it's a very common task we all need when working with this library.

**Comment by stas** [comment]
@patodevilla I'll be making Ransack a soft-dependency in v1.8:
https://github.com/stas/jsonapi.rb/milestone/1

Btw, as mentioned previously with @ekampp, none of the feedback you guys are talking about reached me at `jsonapi.rb` or here.

You can obviously fork and build your own implementations, but I think we should all be realistic about maintaining all this stuff together. Hence this discussion, since orchestrating API needs across multiple gems is... well, intense lol :see_no_evil: 

Personally I've been on the JSONAPI ship for plenty of years, and I've seen plenty of projects stall due to lack of maintenance. :nail_care: 

@patodevilla btw, `jsonapi-deserializable` is unmaintained, good luck using that. That was the main reason I added deserialization support for `jsonapi.rb`:
https://github.com/stas/jsonapi.rb#deserialization

**Comment by jrochkind** [comment]
I wound up here after being surprised that jsonapi-serializer didn't support pagination, even for serializing. It seems like a core serializing feature to me; but maybe there's some way to put the pagination keys in the serialization I haven't figured out yet?

But anyway... it's clear the maintenance resources for _any_ of these gems are dwindling, making it hard to even keep supporting them as they are; combining them to ease support burden seems attractive, but I'm not sure there is currently resources even to do that, or to do maintenance once it's done? 

The current README of this project says "contributions for new features/improvements are welcome only for v3", but the "v3 pull request" has a comment Q: "Is there any tentative timeline for releasing this version? ", A: "to be honest, not really, as I feel like everyone is just happy with v2."   Which suggests that there isn't really any development of new features/improvements happening at all.  The general impression is that this project is receiving very minimal maintenance at present. If merging the two gems would result in a change to that, that seems good, but... unclear that it would.

**Comment by stas** [comment]
@jrochkind I think the topics you're touching are different, there's shipping v3 and maintenance for v2.

I honestly was hoping to see more interest in v3, but there's none or very little. 1-2 times folks raised the question if this project is dead, and my opinion didn't changed: I don't want to change things that are working. And I'm running 2-3 projects which fully depend on the API side on `jsonapi.rb` and this one.

I think the two projects are actually mature enough already. I haven't seen bugs or security issues, so I'm not stressing for releases just to bump numbers. Which to me suggests we're in a good spot. But I'm obviously happy to hear what needs to change. 

Appreciate you stepping in! :bow:

**Comment by jrochkind** [comment]
Thanks for reponse! This might not be the right thread for it, but I guess I'm confused about the stance towards new features/improvements. 


If the README is correct that new features/improvements are not welcome in 2.x/main branch, but it's also true that a 3.x release is not actually currently in the cards... does that mean that jsonapi-serializer is not currently accepting new features at all, for the foreseeable future?

(new features could hypothetically include: support for some things in [jsonapi 1.1 in development](https://jsonapi.org/format/1.1/); better support for serializing pagination links and metadata in jsonapi-serializer; etc)

As far as _this_ thread, if you have decided _not_ to merge the two gems, perhaps close this issue? I am not currently a user of these gems, but just trying to figure out the lay of the land, and finding lots of open discussions etc., leaving me with questions as to current status.

**Comment by stas** [comment]
> but it's also true that a 3.x release is not actually currently in the cards...

Incorrect, the PR describes the current state of things, it's just only me who was pushing for it. The core functionality runs fine, the changes fix and open up the doors for new features which previously were blocked because of a rushed release/incomplete implementation of some specs (full list is in the #141 description).

> Should the README be changed to not suggest new features should be sent to 3.x branch, if that's in fact not a thing that's going to happen?

I really don't advise and support the idea of merging new features into v2. An easy walk through the tasks described in #141 should give a pretty accurate picture of why v2 is so painful to maintain and improve. Take the simple thing like how the serializers are resolved for every relationship, and it's pretty clear what a mess would it be to have backwards compatibility for the old interface :flushed: 



> (new features could hypothetically include: support for some things in [jsonapi 1.1 in development](https://jsonapi.org/format/1.1/); better support for serializing pagination links and metadata in jsonapi-serializer; etc; but when the README says new features are not welcome, I wouldn't ordinarily even bring them up, right?)

Yeah, v1.1 is been in progress for a long time. This could be a good motivating factor to move forward with the release. Tbh, I'd like to see v1.1 spec released first before we start adopting the changes.



> As far as _this_ thread, if you have decided _not_ to merge the two gems, perhaps close this issue? I am not currently a user of these gems, but just trying to figure out the lay of the land, and finding lots of open discussions etc., leaving me with questions as to current status.

Totally, my bad. Had indeed to close it for a while. As to your question re the maintenance... Look, I'll be honest, there's been little feedback and support to move forward with a new version. I know some bigger companies using my stuff, none stood up to support this work (probably because everything works lol, so it's my bad). The other folks who joined this org, have been extremely passive as well. So it's down to me to decide how do we manage the resources here, and not forcing things and approaching things _organically_ proved it's fine like this.

Closing this, and will create a new issue to look into v1.1 spec. Ty for sharing that!

**Comment by jrochkind** [comment]
Thanks!

If i had a new feature to propose or PR, I'm still confused about what is invited though. Not welcome in 2.x, and in 3.x it's unlikely to see the light of day.... Or, maybe I'm not confused, maybe I'm getting the idea that currently new feature requests or PR's are _not_ invited on this project, if one comes up I should not bother trying to describe it or PR it here?

**Comment by jrochkind** [comment]
In particular I am currently thinking about better support for pagination links in serialization... But I am concluding there is no reason to file an issue or PR, it's not invited here;  that is something I should just do myself in local code and/or a fork?

**Comment by stas** [comment]
@jrochkind tbh, maybe just start with opening an issue and not a PR?

I usually try to help/provide support directions on new requests.



> I am currently thinking about better support for pagination links in serialization

If this is part of the v1.0 spec (which I don't remember it to be), I'll be happy to review it. If it's not, it's probably worth looking into the `jsonapi.rb` for that sort of functionality.

This project is really just about getting the spec implemented, hence the lack of interest to merge with `jsonapi.rb` :shrug:

**Comment by jrochkind** [comment]
OK, I'm still stuck on the fact that the README says new features are not welcome in 2.x, but the 3.x branch says there is no current expected release of 3.x. 

So from that I conclude, that new features are not invited... but ok, I will consider  opening an issue and getting more feedback there, since you clearly are responsive! thanks!

Yes, pagination links are part of the 1.0 spec. https://jsonapi.org/format/#fetching-pagination They are supported by your jsonapi.rb.  I don't need all of what's in jsonapi.rb, I really do just want serialization. But better support for pagination links in jsonapi-serializer seems like a good idea to me, instead of having to completely hand-roll it (which I think would be possible?). But if this hasn't come up before, perhaps I'm not totally seeing the big picture.

**Comment by stas** [comment]
> Yes, pagination links are part of the 1.0 spec. https://jsonapi.org/format/#fetching-pagination They are supported by your jsonapi.rb. I don't need all of what's in jsonapi.rb, I really do just want serialization. But better support for pagination links in jsonapi-serializer seems like a good idea to me, instead of having to completely hand-roll it (which I think would be possible?). But if this hasn't come up before, perhaps I'm not totally seeing the big picture.

The spec says:
> A server MAY provide links to traverse a paginated data set (“pagination links”).

It was not implemented because the implementation is specific to how one's integrating the library (you need access to the request object and support for the serializable object type, be that ActiveRecord or other flavor). I'd also argue if that's an actual concern of the "serailization" purpose, which is in fact this gem doing.

Which makes an argument why the pagination is not part of the features. 

My personal opinion is that I already wrote the pagination support (which folks didn't want to merge lol). It's provided with 0 dependencies, full docs and tests and I'd personally not support adding it to this library.

In fact, I'd be open to hear what's stopping you from using `jsonapi.rb`, because _I don't need all of what's in jsonapi.rb_, is not enforced by the library :thinking: 
https://github.com/stas/jsonapi.rb/blob/master/lib/jsonapi/pagination.rb

**Comment by jrochkind** [comment]
OK, no need for opening an issue then, thanks!

This is only meant to support things that are _required_ by the spec, not things allowed with "may"? I'm pretty confused, I don't think that's true, it already supports all sorts of "may" stuff, doesn't it?  But regardless, you don't want to support pagination here, that's your call!

I think I can maybe still add pagination links myself, I'll experiment with it. 

> My personal opinion is that I already wrote the pagination support (which folks didn't want to merge lol). It's provided with 0 dependencies, full docs and tests and I'd personally not support adding it to this library without the rest of stuff.

You're talking about jsonapi.rb here?  I don't know the history, or what folks didn't want to merge them together, but I sense some bitterness, sorry! 

It's ultimately up to you as maintainer whether you want to merge them together, right? Just as it's up to you as maintainer that you don't want to add pagination features to jsonapi-serializer unless you are merging all of jsonapi.rb in!  I don't understand it, but your call!  Like I said, personally do support you merging them together if it makes it easier for you to maintain, because energy to maintain is clearly an issue! 

> In fact, I'd be open to hear what's stopping you from using jsonapi.rb, because I don't need all of what's in jsonapi.rb, is not enforced by the library

OK, I could take another look at jsonapi.rb.  In general, I like the level of abstraction that jsonapi-serializer is at, and more code doing more things I don't understand are more places for bugs. 

I didn't love the Rails patches that jsonapi.rb did, but I guess I could use it without running `JSONAPI::Rails.install!` to skip them!

**Comment by stas** [comment]
> You're talking about jsonapi.rb here? I don't know the history, or what folks didn't want to merge them together, but I sense some bitterness, sorry!

No saltiness, just saying it doesn't make sense for me to merge the pagination when I still need to do legwork on how includes/filters work outside of this gem. We either follow the separation of concerns, or don't :upside_down_face: 

> I didn't love the Rails patches that jsonapi.rb did

I'm not sure what do you mean by this. Appreciate if you can provide a concrete example :pray:

**Comment by jrochkind** [comment]
I guess I just mean I don't love module mix-ins to Rails controllers, with specified methods to override locally, as a way of accessing behavior.

**Comment by stas** [comment]
> I guess I just mean I don't love module mix-ins to Rails controllers, with specified methods to override locally, as a way of accessing behavior.

Sorry ser, but we need to allow the folks to adjust the methods and overwrite potentially the default behavior :smiling_face_with_tear: 

A module or a class will be a minimal requirement for such an implementation. And if I have to chose, it will be a module because then anybody can compose their own class with it :v: 

When I suggested a concrete example, I'm really trying to say I want to understand how a module is a problem for composability in the context of this exact problem/solution/implementation.

**Comment by jrochkind** [comment]
I'm sorry if I was being confrontational or accusatory, I didn't mean to be, but sometimes am accidentally. 

I just don't like module mix-ins to Rails controllers as a pattern, I find that it often results in hard-to-understand , hard-to-debug, potential of conflict, "magic" seeming behavior.  I prefer more explicit APIs where I just call things, instead of what we could call "them calling me" (framework style).  So if I am evaluating two gems, and one uses this pattern and one doesn't, I'm going to prefer the one that doesn't, all things being equal. So this is what makes me reluctant to use jsonapi.rb; if I wanted the features it had enough I might anyway of course. 

I think you disagree with my opinion of module mix-in to a Rails controller. That's fine. Perhaps I am in a minority. I don't need to convince you to agree with me, that's fine. 

I don't think this conversation is helpful to either of us at this point.  

Thank you for your information and responses!

**Comment by stas** [comment]
Thank you @jrochkind, no hard feelings. Honestly just trying to listen and understand here. :hugs:

**Comment by ziggythehamster** [comment]
> I know some bigger companies using my stuff, none stood up to support this work (probably because everything works lol, so it's my bad).

We fall into this bucket for sure. I would love to drop our fork (backporting something from v3 into v2, #178) and also get our other PR (https://github.com/stas/jsonapi.rb/pull/45) back to a mergeable state, but everything works so well in this library that we've let it fall off our radar.

I have yet to dig into v3 to see how different its API is (or if I did, I don't remember), but I wonder if this really should be a more meta-discussion on #141 about what it would take to ship v3 and actually have some period where v2 and v3 are maintained in parallel. Perhaps "I'll only take enhancements for v2 as long as they're also done in v3"?

Re: the current "implement a method and we'll call it" API, I generally prefer it (composition > inheritance). I think it would be slightly better if it were a DSL so it's more explicit, but then you're still adding `include JSONAPI::Whatever` so you can `jsonapi_meta :method_name` which then performs the same function as `def jsonapi_meta` except it's `def method_name`. We don't define this stuff in controllers anyways and instead use concerns, so our pagination implementation lives in one place (we use Kaminari). We use macros to define our controller actions, so we could theoretically configure pagination explicitly there, but that seems less ideal. Maybe it would be better to have the current API except you could configure the library to use a separate object from the controller? But then where does it live within app? app/jsonapi/controllers/? And then now I'm bikeshedding this - it works how it is right now and it's quite fast and efficient, and I only had to care about how it works once :).

**Comment by jrochkind** [comment]
Adding behavior to a controller with a module mix-in is definitely inheritance, not composition!  (See for instance https://cek.io/blog/2014/05/26/poodr-ch-7/ for some discussion) But whatever works for you works for you!

**Comment by ziggythehamster** [comment]
From the followup post:

> Composition allows you to combine small parts to create more complex objects such that the whole becomes more than the sum of its parts. Composed objects tend to consist of simple, discrete entities that can easily be rearranged into new combinations. These simple objects are easy to understand, reuse, and test, but because they combine into a more complicated whole, the operation of the bigger application may not be as easy to understand as that of the individual parts.

This tends to be how we use mixins/concerns. e.g., our pagination concern adds this DSL method to controllers which request it:

```ruby
 api_pagination policy: :required do
   default_page_size       50
   maximum_page_size       75
   optional_with_params    %i[one two three]
   unpermitted_with_params %i[foo bar]
 end
```

A controller could instead not set this and a default policy is applied. Under the hood, methods are added to the controller that get picked up by this library and then dispatch pagination to Kaminari, but controller authors need not concern themselves with that - they include this (or inherit from a class which includes it) and that concern adds pagination-related methods that are technically defined on the controller but not part of the API you're normally expected to execute. It would perhaps be better if these were defined on a class inside the class that is private to the controller but only if you intend on defining methods like `jsonapi_pagination` or `configure_and_validate_pagination` and you wouldn't.
